compiler/qsc_frontend/src/compile/preprocess.rs

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use core::str::FromStr;
+use qsc_ast::{
+    ast::{Attr, ExprKind, Namespace},
+    mut_visit::MutVisitor,
+};
+
+use super::Target;
+
+pub(super) struct Conditional {
+    pub(super) target: Target,
+}
+
+impl MutVisitor for Conditional {
+    fn visit_namespace(&mut self, namespace: &mut Namespace) {
+        namespace.items = namespace
+            .items
+            .iter()
+            .filter_map(|item| {
+                if matches_target(&item.attrs, self.target) {
+                    Some(item.clone())
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>()
+            .into_boxed_slice();
+    }
+}
+
+fn matches_target(attrs: &[Box<Attr>], target: Target) -> bool {
+    attrs.iter().all(|attr| {
+        if attr.name.name.as_ref() == "TargetProfile" {
```

Comment:
We could consider naming this attribute just `Target`. It is more general, and I believe removing "Profile" for it would be less ambiguous to people used to other languages since that is the terminology used in most of them.

Reply:
It also keeps it more flexible because we could refer to QIR profile or target feature/capabilities (or any other thing that we would need to add) as properties of a `Target`.

Reply:
The name was definitely something we found could use some extra thought. I agree that `Target` is more general while still being specific to the usage I'm looking for, so I can make that change.